### ISSUE TITLE
index_merge returns wrong result 

### ISSUE THREAD

**Issue by XuHuaiyu** [pingcap/tidb#23569]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
``` sql
set @@tidb_enable_index_merge=true;
create table tt(`id` bigint(20) NOT NULL,`create_time` bigint(20) NOT NULL DEFAULT '0' ,`driver` varchar(64), PRIMARY KEY (`id`,`create_time`))
PARTITION BY RANGE ( `create_time` ) (
  PARTITION `p201901` VALUES LESS THAN (1577808000),
  PARTITION `p202001` VALUES LESS THAN (1585670400),
  PARTITION `p202002` VALUES LESS THAN (1593532800),
  PARTITION `p202003` VALUES LESS THAN (1601481600),
  PARTITION `p202004` VALUES LESS THAN (1609430400),
  PARTITION `p202101` VALUES LESS THAN (1617206400),
  PARTITION `p202102` VALUES LESS THAN (1625068800),
  PARTITION `p202103` VALUES LESS THAN (1633017600),
  PARTITION `p202104` VALUES LESS THAN (1640966400),
  PARTITION `p202201` VALUES LESS THAN (1648742400),
  PARTITION `p202202` VALUES LESS THAN (1656604800),
  PARTITION `p202203` VALUES LESS THAN (1664553600),
  PARTITION `p202204` VALUES LESS THAN (1672502400),
  PARTITION `p202301` VALUES LESS THAN (1680278400)
);

insert tt value(1, 1577807000, 'jack'), (2, 1577809000, 'mike'), (3, 1585670500, 'right'), (4, 1601481500, 'hello');

select count(*) from tt partition(p202003) where _tidb_rowid is null or (_tidb_rowid>=1 and _tidb_rowid<100);
```

### 2. What did you expect to see? (Required)
``` sql
+----------+
| count(*) |
+----------+
|        1 |
+----------+
```

### 3. What did you see instead (Required)
``` sql
``` sql
+----------+
| count(*) |
+----------+
|        0 |
+----------+
```
```

### 4. What is your TiDB version? (Required)

release-4.0: 11a9254bec522c1991033a6098f6dccadc6eb7c1
release-5.0 and master do not have this bug

**Comment by XuHuaiyu** [comment]
This bug has been fixed in #23572 and there are 2 LGTM already.
Since we plan to not release it in 4.0.12, I'll remove the `critical` and `bug` label temporarily to avoid been moved into the bug jail.

**Comment by mjonss** [comment]
This was fixed by #23572 and is works as expected in v4.0.16 (latest 4.0 patch version).
